### PR TITLE
Infer linted values that resolve to dbutils.widgets.get

### LINKED ISSUE #1880: [FEATURE]: infer linted values that resolve to dbutils.widgets.get()
`dbutils.widgets.get(X)` can retrieve values supplied from Job's named parameters. These values may contain random data, including storage location prefixes and table names:

```
selected_table = dbutils.widgets.get("selected_table")
query = f"select * from some_schema.{selected_table}"
spark.sql(query").collect()
```

As of https://github.com/databrickslabs/ucx/commit/de4fa7fddcb9ee7e58111c84ae852e0672ee8e07 we fetch named parameters into CurrentSessionState:

https://github.com/databrickslabs/ucx/blob/de4fa7fddcb9ee7e58111c84ae852e0672ee8e07/src/databricks/labs/ucx/source_code/base.py#L131-L147

# In scope:
- Resolve every `dbutils.widgets.get("XXXXXXXX")` into `CurrentSessionState#named_parameters.get("XXXXXXXX")`
- [stretch-goal] `dbutils.widgets.getAll()` to resolve into `CurrentSessionState#named_parameters` value

# Out of scope:
- parsing of any other methods of `dbutils.widgets` than `get(str) -> str` and `getAll() -> dict[str,str]`.

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
When linting `dbutils.widget.get(...)` or `dbutils.widget.getAll()`, use named parameters from current session state to resolve the value

### Linked issues
Resolves #1880

### CODE REVIEW

File: src/databricks/labs/ucx/contexts/application.py [hunk 1]
```diff
@@ -375,6 +376,10 @@ def path_lookup(self):
         # TODO find a solution to enable a different cwd per job/task (maybe it's not necessary or possible?)
         return PathLookup.from_sys_path(Path.cwd())
 
+    @cached_property
+    def session_state(self):
```
Comment (nfx) [inline src/databricks/labs/ucx/contexts/application.py:380]: current session state has to be created per-job-task in `LinterContext`, not per global context. revert.

File: src/databricks/labs/ucx/contexts/application.py [hunk 2]
```diff
@@ -393,7 +398,9 @@ def file_resolver(self):
 
     @cached_property
     def dependency_resolver(self):
-        return DependencyResolver(self.pip_resolver, self.notebook_resolver, self.file_resolver, self.path_lookup)
```
Comment (nfx) [inline src/databricks/labs/ucx/contexts/application.py:396]: revert this change

File: src/databricks/labs/ucx/source_code/base.py [hunk 3]
```diff
@@ -127,7 +127,7 @@ class Convention(Advice):
 
 class Linter:
     @abstractmethod
-    def lint(self, code: str) -> Iterable[Advice]: ...
+    def lint(self, code: str, session_state: CurrentSessionState) -> Iterable[Advice]: ...
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/base.py:130]: why do we need to change the interface? why it can't be injected via constructor only to those linters than need to be aware of the current session state? revert.

File: src/databricks/labs/ucx/source_code/base.py [hunk 4]
```diff
@@ -167,6 +167,6 @@ class SequentialLinter(Linter):
     def __init__(self, linters: list[Linter]):
         self._linters = linters
 
-    def lint(self, code: str) -> Iterable[Advice]:
+    def lint(self, code: str, session_state: CurrentSessionState) -> Iterable[Advice]:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/base.py:170]: revert

File: src/databricks/labs/ucx/source_code/graph.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 from astroid import ImportFrom, NodeNG  # type: ignore
 
-from databricks.labs.ucx.source_code.base import Advisory
+from databricks.labs.ucx.source_code.base import Advisory, CurrentSessionState
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:11]: revert changes to this file.

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 6]
```diff
@@ -130,28 +142,35 @@ def _get_attribute_value(cls, node: Attribute):
...
         return None
 
-    def infer_values(self) -> Iterable[InferredValue]:
+    def infer_values(self, state: CurrentSessionState | None = None) -> Iterable[InferredValue]:
+        if state is not None:
+            self.contextualize(state)
         for inferred_atoms in self._infer_values():
             yield InferredValue(inferred_atoms)
 
+    def contextualize(self, state: CurrentSessionState):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:151]:
```suggestion
    def _contextualize(self, state: CurrentSessionState):
```
make private

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 7]
```diff
@@ -130,28 +142,35 @@ def _get_attribute_value(cls, node: Attribute):
...
+        calls = Tree(self.root).locate(Call, [("get", Attribute), ("widgets", Attribute), ("dbutils", Name)])
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:152]: would it also work for

```
x = dbutils.widgets
x.get(Y)
```
?

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 8]
```diff
@@ -170,6 +189,51 @@ def do_infer_values(self):
         return self._infer_values()
 
 
+class _ContextualCall(NodeNG):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:192]:
```suggestion
class _GetWidgetValue(NodeNG):
```
make name more explicit

File: src/databricks/labs/ucx/source_code/linters/python_ast.py [hunk 9]
```diff
@@ -130,28 +142,35 @@ def _get_attribute_value(cls, node: Attribute):
             logger.debug(f"Missing handler for {name}")
         return None
 
-    def infer_values(self) -> Iterable[InferredValue]:
+    def infer_values(self, state: CurrentSessionState | None = None) -> Iterable[InferredValue]:
+        if state is not None:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/linters/python_ast.py:146]: check for the presence of named parameters here

**Review by nfx — CHANGES_REQUESTED** [review]
don't change the `lint(str) -> Iterable[Advice]` interface. perhaps you want to do `lint(Tree) -> Iterable[Advice]`, right?

File: src/databricks/labs/ucx/source_code/linters/python_ast.py (hunk 7 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/python_ast.py:152]: No, same as #1864

File: src/databricks/labs/ucx/contexts/application.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/contexts/application.py:380]: done

File: src/databricks/labs/ucx/contexts/application.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/contexts/application.py:396]: done

File: src/databricks/labs/ucx/source_code/base.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/base.py:170]: done

File: src/databricks/labs/ucx/source_code/linters/python_ast.py (hunk 6 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/python_ast.py:151]: done

File: src/databricks/labs/ucx/source_code/linters/python_ast.py (hunk 8 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/python_ast.py:192]: done

File: src/databricks/labs/ucx/source_code/linters/python_ast.py (hunk 9 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/linters/python_ast.py:146]: done

File: src/databricks/labs/ucx/source_code/graph.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:11]: we need a session state to compute dependencies that rely on named parameters, so not sure it's possible to not change this file

**Comment by ericvergnaud** [comment]
> perhaps you want to do `lint(Tree) -> Iterable[Advice]`, right?
Possibly, but not sure how that would help convey a state...

**Review by nfx — CHANGES_REQUESTED** [review]
Session state is initialized per LinterContext (eg, per job task). This enables us to add parallel linting for all jobs.

File: src/databricks/labs/ucx/source_code/graph.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:11]: Sure, but what do you propose?

Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/graph.py:11]: And is it correct to NOT share context between files that belong to the same graph ? IIUC DB executes notebooks synchronously so parallelizing linting could introduce all sorts of issues ?

**Review by nfx — APPROVED** [review]
Lgtm
